File: Data Structures & Algorithms/remove-linked-list-elements/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode curr = head; 
        List<Integer> list = new ArrayList<>(); 

        while (curr != null) {
            list.add(curr.val);
            curr = curr.next; 
        }

        //System.out.println(list);
        for (int i = list.size() - 1; i >= 0; i--) {
            if (list.get(i) == val) {
                list.remove(i);
            }
        }

        if (list.size() == 0) {
            return null;
        }

        ListNode newList = new ListNode(list.get(0));
        ListNode curr2 = newList; 
        for (int i = 1; i < list.size(); i++) {
            ListNode newNode = new ListNode(list.get(i)); 
            curr2.next = newNode; 
            curr2 = curr2.next;

        }
        return newList;
    }
}

File: Data Structures & Algorithms/singlyLinkedList/submission-0.py
class ListNode:
    def __init__(self, val, next_node=None):
        self.val = val
        self.next = next_node

class LinkedList:
    
    def __init__(self):
        #head 
        self.head = ListNode(-1)
        self.tail = self.head
      
    
    def get(self, index: int) -> int:
        curr = self.head.next
        i = 0
        #need to get length of the list first:
        while curr:
            if i == index:
                return curr.val
            i += 1
            curr = curr.next
        return -1
        
        
            
        
    def insertHead(self, val: int) -> None:
        # first make new node:
        new_node = ListNode(val)
        new_node.next = self.head.next
        self.head.next = new_node

        #check if the list was empty:
        if new_node.next == None:
            self.tail = new_node # if list is empty, new node becomes both head and tail

    def insertTail(self, val: int) -> None:
        new_node = ListNode(val)
        self.tail.next = new_node
        self.tail = self.tail.next
        

    def remove(self, index: int) -> bool:
        cur = self.head
        i = 0

        while i < index and cur:
            i += 1
            cur = cur.next
        
        #remove the node ahead of cur:
        #check if they are not none
        if cur and cur.next:
            # check if cur.next is tail (we are at the end of the list)
            # so we will need to also update the tail

            if cur.next == self.tail:
                self.tail = cur
            
            cur.next = cur.next.next
            return True
        return False
        

    def getValues(self) -> List[int]:
        arr = []
        curr = self.head.next
        while (curr != None):
            arr.append(curr.val)
            curr = curr.next
        return arr
